doc/reference/trd-syscalls.md

```diff
@@ -264,26 +264,51 @@ call to invoke. The kernel invokes callbacks only in response to Yield
 system calls.  This form of very limited preemption allows userspace
 to manage concurrent access to its variables.
 
-There are two Yield system calls:
+There are three Yield system calls:
   - `yield-wait`
   - `yield-no-wait`
+  - `yield-exit`
 
 The first call, `yield-wait`, blocks until a callback executes. This is the
 only blocking system call in Tock. It is commonly used to provide a blocking
 I/O interface to userspace. A userspace library starts a long-running operation
 that has a callback, then calls `yield-wait` to wait for a callback. When the
 `yield-wait` returns, the process checks if the resuming callback was the one
-it was expecting, and if not calls `yield-wait` again. The `yield-wait` system
-call always returns the `Success` variant.
+it was expecting, and if not calls `yield-wait` again. 
 
-The second call, 'yield-no-wait', executes a single callback if any is
-pending.  If no callbacks are pending it returns immediately. The
-`yield-no-wait` system call returns the `Success` variant if a
-callback executed and `Failure` variant if no callback executed.
+The second call, `yield-no-wait`, executes a single callback if any is pending.
+If no callbacks are pending it returns immediately. 
 
-The return variants for Yield system calls are `Success` and
-`Failure`.
+The third call, `yield-exit` never returns and executes no callbacks. It
+causes the process to terminate. The kernel can reclaim any resources held by
+a process after it invokes `yield-exit`.
 
+The register arguments for Subscribe system calls are as follows. The registers
+r0-r3 correspond to r0-r3 on CortexM and a0-a3 on RISC-V.
+
+| Argument               | Register |
+|------------------------|----------|
+| Yield identifer        | r0       |
+| No wait field          | r1       |
+
+
+The yield identifier specifies which call is invoked.
+
+| System call     | Yield identifier value |
+|-----------------|------------------------|
+| yield-no-wait   |                      0 |
+| yield-wait      |                      1 |
+| yield-exit      |                      2 |
+
+
+The no wait field is only used by `yield-no-wait`. It contains the
```

Comment:
I would still like to see processes be able to include an argument when they call `exit`, and have a convention where the process can indicate to the kernel if it would like to be restarted.

Reply:
Regarding the exit argument and restart policy, could it make sense to have a separate syscall for that (essentially a "process control system call")? That way the process could set a return value / set the restart policy manually using this command, and then call yield-exit (with the usual yield semantics, but just stopping execution indefinitely)?

Reply:
Also since Linux came up on Friday, there is an explicit exit syscall (https://man7.org/linux/man-pages/man2/_exit.2.html) in Linux which just [exits](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/_exit.c.html#_exit), and is separate from [yield](https://man7.org/linux/man-pages/man2/sched_yield.2.html). I still think yielding and exiting are separate, but `yield-exit` is pretty specific and clear. But even though `exit()` in Linux does do other things, it does eventually call the exit syscall.

Reply:
+1 to exit having a return value.

Reply:
Exit has a return value or an argument?

If exit has an argument, we should specify what the argument means. We need to specify its values and their semantics.

I have a different proposal: there should be `yield-exit` and `yield-exit-restart`. We can extend different exit cases by adding more system calls.

Reply:
>  have a different proposal: there should be yield-exit and yield-exit-restart. We can extend different exit cases by 
> adding more system calls.

Wouldn't adding more syscalls for every exit condition complicate the interface a bit?  Adding an "exit status" argument could make it more similar to Linux's `void _exit(int status)` call.

Reply:
Now that we have two exit cases, I think I'm on board with Brad's suggestion. We should have a yield with two variants (-wait and -no-wait), and an exit with two variants (-terminate and -restart).

Reply:
I've factored out exit from yield. I think they are semantically different: exit never returns and doesn't execute syscalls. If we need to expand the set of cases for either we'll want them distinct, in terms of shared functionality. Otherwise I'm worried about a fragmented identifier space (e.g., 0, 1, 4, 5 are yields, 2, 3, 6 are exits, etc.).